url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/Simulation.java

```diff
@@ -0,0 +1,1042 @@
+					//		  				cycleMaxPower = 0;
+					//		  			}
+					m.addAggregatedTickResult(tick, 
+							sumP * mcrunsRatio, 
+							sumQ * mcrunsRatio, 
+							DBResults.COL_AGGRRESULTS);
+					tick++;
+					if(tick % Constants.MIN_IN_HOUR == 0) {
+						m.addAggregatedTickResult((tick/Constants.MIN_IN_HOUR), 
+								(avgPPowerPerHour/Constants.MIN_IN_HOUR) * mcrunsRatio, 
+								(avgQPowerPerHour/Constants.MIN_IN_HOUR) * mcrunsRatio, 
+								DBResults.COL_AGGRRESULTS_HOURLY);
+						m.addAggregatedTickResult((tick/Constants.MIN_IN_HOUR), 
+								(avgPPowerPerHour) * mcrunsRatio, 
+								(avgQPowerPerHour) * mcrunsRatio, 
+								DBResults.COL_AGGRRESULTS_HOURLY_EN);
+						avgPPowerPerHour = 0;
+						avgQPowerPerHour = 0;
+						counter = 0;
+						for(Installation installation: installations) {
+							m.addTickResultForInstallation((tick/Constants.MIN_IN_HOUR), 
+									installation.getId(),
+									(avgPPowerPerHourPerInst[counter]/Constants.MIN_IN_HOUR) * mcrunsRatio, 
+									(avgQPowerPerHourPerInst[counter]/Constants.MIN_IN_HOUR) * mcrunsRatio, 
+									DBResults.COL_INSTRESULTS_HOURLY);
+							m.addTickResultForInstallation((tick/Constants.MIN_IN_HOUR), 
+									installation.getId(),
+									(avgPPowerPerHourPerInst[counter]) * mcrunsRatio, 
+									(avgQPowerPerHourPerInst[counter]) * mcrunsRatio, 
+									DBResults.COL_INSTRESULTS_HOURLY_EN);
+							avgPPowerPerHourPerInst[counter] = 0;
+							avgQPowerPerHourPerInst[counter] = 0;
+							counter++;
+						}
+					}
+					mccount++;
+					//		  			percentage = (int)(0.75 * mccount * 100.0 / (mcruns * endTick));
+					//		  			System.out.println("Percentage: " + percentage + " - " + mccount);
+					//		  			objRun.put("percentage", 25 + percentage);
+					//		  	  		DBConn.getConn().getCollection(MongoRuns.COL_RUNS).update(query, objRun);
+				}
+				for(Installation installation: installations) {
+					installation.updateCost(tick); // update the rest of the energy
+					m.addKPIs(installation.getId(), 
+							installation.getMaxPower() * mcrunsRatio, 
+							installation.getAvgPower() * mcrunsRatio, 
+							installation.getEnergy() * mcrunsRatio, 
+							installation.getCost() * mcrunsRatio,
+							installation.getEnergy() * co2 * mcrunsRatio);
+					installation.addAppliancesKPIs(m, mcrunsRatio, co2);
+					installation.addActivitiesKPIs(m, mcrunsRatio, co2);
+				}
+				cost = totalInstCost();
+				m.addKPIs(DBResults.AGGR, 
+						maxPower * mcrunsRatio, 
+						avgPower * mcrunsRatio, 
+						energy * mcrunsRatio, 
+						cost * mcrunsRatio,
+						energy * co2 * mcrunsRatio);
+				if(i+1 != mcruns) setup(true);
+
+			}
+			// Write installation results to csv file
+			if (!resources_path.endsWith("/"))
+				resources_path += "/";
+			String filename = resources_path + dbname + ".csv";
+			//  			System.out.println(filename);
+			File csvFile = new File(filename);
+			FileWriter fw = new FileWriter(csvFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fw*. The resource is referenced at lines 740, 768, 770. The resource is closed at line 771. There are other execution paths that do not contain closure statements, for example, when OutputStreamWriter.flush throws an exception. Ensure *fw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).